src/Game.js

```diff
@@ -4,155 +4,126 @@ function isValid(data, size, x, y) {
     return x >= 0 && x < size && y >= 0 && y < size && Number.isInteger(data[y * size + x]);
 }
 
-export function RookMoves(data, size, x, y) {
+function parseMove(dx, dy, length, constraints, data, size, x, y) {
     let moves = [];
-    for (let yi = y - 1; yi >= 0; yi--) {
-        if (Number.isInteger(data[yi * size + x])) moves.push(yi * size + x);
-        else break;
-    }
-    for (let yi = y + 1; yi < size; yi++) {
-        if (Number.isInteger(data[yi * size + x])) moves.push(yi * size + x);
-        else break;
-    }
-    for (let xi = x - 1; xi >= 0; xi--) {
-        if (Number.isInteger(data[y * size + xi])) moves.push(y * size + xi);
-        else break;
+    let orientation = [];
+    const directions = [
+        [-dx, -dy], // Forward
+        [dx, dy], // Backward
+        [-dy, dx], // Right
+        [dy, -dx] // Left
+    ];
+    for (let d of [-1, 1]) { // For pieces like knights, we need to reverse the X for each direction
+        for (let rd in directions) {
+            if ((constraints & (2 ** rd)) === 0) {
+                continue;
+            }
+            const nrd = [directions[rd][0], directions[rd][1] * d];
+            if (orientation.every(x => x[0] !== nrd[0] || x[1] !== nrd[1])) orientation.push(nrd);
+        }
     }
-    for (let xi = x + 1; xi < size; xi++) {
-        if (Number.isInteger(data[y * size + xi])) moves.push(y * size + xi);
-        else break;
+    for (let o of orientation) {
```

Comment:
not a big deal but could destructure here if you want
`for (const [yi, xi] of o)`

Reply:
I'll edit that, thx